services/remotecluster/send.go

```diff
@@ -14,95 +14,120 @@ import (
 // BufferFullError if the message cannot be enqueued before the timeout. A background context will block indefinitely.
 //
 // An optional callback can be provided that receives the success or fail result of sending to each remote cluster.
-// Success or fail is regarding message delivery only. If a callback is provided it should return quickly.
-func (rcs *Service) SendOutgoingMsg(ctx context.Context, msg model.RemoteClusterMsg, f SendResultFunc) error {
-	task := sendTask{
-		msg: msg,
-		f:   f,
+// Success or fail is regarding message delivery only.  If a callback is provided it should return quickly.
+func (rcs *Service) BroadcastMsg(ctx context.Context, msg model.RemoteClusterMsg, f SendResultFunc) error {
+	// get list of interested remotes.
+	filter := model.RemoteClusterQueryFilter{
+		Topic: msg.Topic,
+	}
+	list, err := rcs.server.GetStore().RemoteCluster().GetAll(filter)
+	if err != nil {
+		return err
+	}
+
+	errs := merror.New()
+
+	for _, rc := range list {
+		if err := rcs.SendMsg(ctx, msg, rc, f); err != nil {
+			errs.Append(err)
+		}
 	}
+	return errs.ErrorOrNil()
+}
 
+// SendMsg asynchronously sends a message to a remote cluster.
+//
+// `ctx` determines behaviour when the outbound queue is full. A timeout or deadline context will return a
+// BufferFullError if the message cannot be enqueued before the timeout. A background context will block indefinitely.
+//
+// An optional callback can be provided that receives the success or fail result of sending to the remote cluster.
+// Success or fail is regarding message delivery only.  If a callback is provided it should return quickly.
+func (rcs *Service) SendMsg(ctx context.Context, msg model.RemoteClusterMsg, rc *model.RemoteCluster, f SendResultFunc) error {
 	if ctx == nil {
 		ctx = context.Background()
 	}
 
+	task := sendTask{
+		rc:  rc,
+		msg: msg,
+		f:   f,
+	}
+
 	select {
 	case rcs.send <- task:
 		return nil
 	case <-ctx.Done():
-		return fmt.Errorf("Could not enqueue outbound remote cluster message; queue full or timeout. Queue %d/%d. Err: %w", len(rcs.send), cap(rcs.send), ctx.Err())
+		return NewBufferFullError(cap(rcs.send))
 	}
+
 }
 
 func (rcs *Service) sendLoop(done chan struct{}) {
-	// create thread pool for concurrent message sending.
-	for i := 0; i < MaxConcurrentSends; i++ {
-		go func() {
-			for {
-				select {
-				case task := <-rcs.send:
-					rcs.sendMsg(task)
-				case <-done:
-					return
-				}
-			}
-		}()
+	for {
+		select {
+		case task := <-rcs.send:
+			rcs.sendMsg(task)
+		case <-done:
+			return
+		}
 	}
 }
 
 func (rcs *Service) sendMsg(task sendTask) {
-	// get list of interested remotes.
-	list, err := rcs.server.GetStore().RemoteCluster().GetByTopic(task.msg.Topic)
-	if err != nil && task.f != nil {
-		task.f(task.msg, nil, nil, fmt.Errorf("could not send msg to remote: %w", err))
-		return
+	// Ensure a panic from the callback doesn't exit the pool thread.
+	defer func() {
```

Comment:
If we panic during this operation, we have already popped that task from the channel, which means that we've lost that task forever right?

Reply:
A panic here is going to come from the callback unless there is a bug in the standard lib's http package.  Therefore the message has already been sent to the remote, and it's just the callback that isn't going to get notified.  The only reason a panic handler is here is that code is getting injected into this package via a callback.